libraries/Microsoft.Bot.Builder.Azure/CosmosDbStorage.cs

```diff
@@ -0,0 +1,269 @@
+        ///     More information at https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.connectionpolicy?view=azure-dotnet
+        /// </param>
+        public CosmosDbStorage(Uri serviceEndpoint, string authKey, string databaseId, string collectionId, Action<ConnectionPolicy> connectionPolicyConfigurator = null)
+        {
+            if (serviceEndpoint == null)
+            {
+                throw new ArgumentNullException(nameof(serviceEndpoint), "Service EndPoint for CosmosDB is required.");
+            }
+
+            if (string.IsNullOrEmpty(authKey))
+            {
+                throw new ArgumentException("AuthKey for CosmosDB is required.", nameof(authKey));
+            }
+
+            if (string.IsNullOrEmpty(databaseId))
+            {
+                throw new ArgumentException("DatabaseId is required.", nameof(databaseId));
+            }
+
+            if (string.IsNullOrEmpty(collectionId))
+            {
+                throw new ArgumentException("CollectionId is required.", nameof(collectionId));
+            }
+
+            _databaseId = databaseId;
+            _collectionId = collectionId;
+
+            // Inject BotBuilder version to CosmosDB Requests
+            var version = GetType().Assembly.GetName().Version;
+            var connectionPolicy = new ConnectionPolicy()
+            {
+                UserAgentSuffix = $"Microsoft-BotFramework {version}"
+            };
+
+            // Invoke CollectionPolicy delegate to further customize settings
+            connectionPolicyConfigurator?.Invoke(connectionPolicy);
+            _client = new DocumentClient(serviceEndpoint, authKey, connectionPolicy);
+
+            // Delayed Database and Collection creation if they do not exist. This is a thread-safe operation.
+            _collectionReady = new Lazy<string>(() =>
+            {
+                _client.CreateDatabaseIfNotExistsAsync(new Database { Id = _databaseId })
+                    .Wait();
+
+                var task = _client.CreateDocumentCollectionIfNotExistsAsync(UriFactory.CreateDatabaseUri(_databaseId), new DocumentCollection { Id = _collectionId });
+                task.Wait();
+
+                return task.Result.Resource.SelfLink;
+            });
+        }
+
+        /// <summary>
+        /// Deletes the specified keys from storage.
+        /// </summary>
+        /// <param name="keys">List of keys to remove.</param>
+        /// <returns></returns>
+        public async Task Delete(params string[] keys)
+        {
+            if (keys.Length == 0) return;
+
+            // Ensure collection exists
+            var collectionLink = _collectionReady.Value;
+
+            // Parallelize deletion
+            var tasks = keys.Select(key =>
+            {
+                return _client.DeleteDocumentAsync(UriFactory.CreateDocumentUri(_databaseId, _collectionId, SanitizeKey(key)));
+            });
```

Comment:
I like this, I normally new up a List, but this is way more succinct

Reply:
Yeah, love this pattern. Technically it can be even more succinct than this by using an expression:

```CSharp
var deleteTasks = keys.Select(key => 
    _client.DeleteDocumentAsync(UriFactory.CreateDocumentUri(_databaseId, _collectionId, SanitizeKey(key)))); 
```